apps/arweave/src/ar_peer_intervals.erl

```diff
@@ -26,79 +31,187 @@
 -define(GET_SYNC_RECORD_BATCH_SIZE, 2).
 -define(GET_SYNC_RECORD_COOLDOWN_MS, 60 * 1000).
 -define(GET_SYNC_RECORD_RPM_KEY, data_sync_record).
+-define(GET_FOOTPRINT_RECORD_RPM_KEY, footprints).
+-define(GET_FOOTPRINT_RECORD_COOLDOWN_MS, 60 * 1000).
 -define(GET_SYNC_RECORD_PATH, [<<"data_sync_record">>]).
+-define(GET_FOOTPRINT_RECORD_PATH, [<<"footprints">>]).
 
 %% The number of the release adding support for the
 %% GET /data_sync_record/[start]/[end]/[limit] endpoint.
 -define(GET_SYNC_RECORD_RIGHT_BOUND_SUPPORT_RELEASE, 83).
 
+%% The number of the release adding support for the
+%% GET /footprints/[partition]/[footprint] endpoint.
+-define(GET_FOOTPRINT_RECORD_SUPPORT_RELEASE, 87).
+
 %%%===================================================================
 %%% Public interface.
 %%%===================================================================
 
-fetch(Start, End, StoreID) when Start >= End ->
+fetch(Start, End, StoreID, normal) when Start >= End ->
 	?LOG_DEBUG([{event, fetch_peer_intervals_end},
 			{store_id, StoreID},
 			{range_start, Start},
-			{range_end, End}]),
-	gen_server:cast(ar_data_sync:name(StoreID), {collect_peer_intervals, Start, End});
-fetch(Start, End, StoreID) ->
+			{range_end, End},
+			{type, normal}]),
+	gen_server:cast(ar_data_sync:name(StoreID), {collect_peer_intervals, Start, End, normal});
+fetch(Start, End, StoreID, normal) ->
 	Parent = ar_data_sync:name(StoreID),
 	spawn_link(fun() ->
-		try
-			End2 = min(Start + ?QUERY_RANGE_STEP_SIZE, End),
-			UnsyncedIntervals = get_unsynced_intervals(Start, End2, StoreID),
-
-			Bucket = Start div ?NETWORK_DATA_BUCKET_SIZE,
-			{ok, Config} = application:get_env(arweave, config),
-			AllPeers =
-				case Config#config.sync_from_local_peers_only of
-					true ->
-						Config#config.local_peers;
-					false ->
-						ar_data_discovery:get_bucket_peers(Bucket)
-				end,
-			HotPeers = [
-				Peer || Peer <- AllPeers,
-				not ar_rate_limiter:is_on_cooldown(Peer, ?GET_SYNC_RECORD_RPM_KEY) andalso
-				not ar_rate_limiter:is_throttled(Peer, ?GET_SYNC_RECORD_PATH)
-			],
-			Peers = ar_data_discovery:pick_peers(HotPeers, ?QUERY_BEST_PEERS_COUNT),
-
-			case length(Peers) < ?QUERY_BEST_PEERS_COUNT of
+		{End2, EnqueueIntervals} = do_fetch(Start, End, StoreID, normal),
+		gen_server:cast(Parent, {enqueue_intervals, EnqueueIntervals}),
+		gen_server:cast(Parent, {collect_peer_intervals, End2, End, normal})
+	end);
+
+fetch(Start, End, StoreID, footprint) when Start >= End ->
```

Comment:
can this be merged with the `normal` code? Looks like they are the same, only difference is the `Type` value?